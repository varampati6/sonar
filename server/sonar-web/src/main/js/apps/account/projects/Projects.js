/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import ProjectCard from './ProjectCard';
import ListFooter from '../../../components/controls/ListFooter';
import { projectsListType } from './propTypes';
import { translate } from '../../../helpers/l10n';

export default class Projects extends React.PureComponent {
  static propTypes = {
    projects: projectsListType.isRequired,
    total: React.PropTypes.number.isRequired,
    loading: React.PropTypes.bool.isRequired,
    loadMore: React.PropTypes.func.isRequired
  };

  render() {
    const { projects } = this.props;

    return (
      <div id="account-projects">
        {projects.length === 0
          ? <div className="js-no-results">
              {translate('my_account.projects.no_results')}
            </div>
          : <p>
              {translate('my_account.projects.description')}
            </p>}

        {projects.length > 0 &&
          <ul className="account-projects-list">
            {projects.map(project => (
              <li key={project.key}>
                <ProjectCard project={project} />
              </li>
            ))}
          </ul>}

        {projects.length > 0 &&
          <ListFooter
            count={projects.length}
            total={this.props.total}
            ready={!this.props.loading}
            loadMore={this.props.loadMore}
          />}
      </div>
    );
  }
}
