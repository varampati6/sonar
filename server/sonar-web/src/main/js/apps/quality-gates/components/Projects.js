/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React, { Component } from 'react';
import ProjectsView from '../views/gate-projects-view';

export default class Projects extends Component {
  componentDidMount() {
    this.renderView();
  }

  componentWillUpdate() {
    this.destroyView();
  }

  componentDidUpdate() {
    this.renderView();
  }

  componentWillUnmount() {
    this.destroyView();
  }

  destroyView() {
    if (this.projectsView) {
      this.projectsView.destroy();
    }
  }

  renderView() {
    const { qualityGate, edit } = this.props;

    this.projectsView = new ProjectsView({
      qualityGate,
      edit,
      container: this.refs.container
    });
    this.projectsView.render();
  }

  render() {
    return <div ref="container" />;
  }
}
