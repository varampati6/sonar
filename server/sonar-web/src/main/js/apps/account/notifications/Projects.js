/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import Select from 'react-select';
import { connect } from 'react-redux';
import { differenceBy } from 'lodash';
import ProjectNotifications from './ProjectNotifications';
import Organization from '../../../components/shared/Organization';
import { translate } from '../../../helpers/l10n';
import { getSuggestions } from '../../../api/components';
import { getProjectsWithNotifications } from '../../../store/rootReducer';

type Props = {
  projects: Array<{
    key: string,
    name: string
  }>
};

type State = {
  addedProjects: Array<{
    key: string,
    name: string
  }>
};

class Projects extends React.PureComponent {
  props: Props;

  state: State = {
    addedProjects: []
  };

  componentWillReceiveProps(nextProps: Props) {
    // remove all projects from `this.state.addedProjects`
    // that already exist in `nextProps.projects`
    const nextAddedProjects = differenceBy(
      this.state.addedProjects,
      nextProps.projects,
      project => project.key
    );

    if (nextAddedProjects.length !== this.state.addedProjects) {
      this.setState({ addedProjects: nextAddedProjects });
    }
  }

  loadOptions = (query, cb) => {
    if (query.length < 2) {
      cb(null, { options: [] });
      return;
    }

    getSuggestions(query)
      .then(r => {
        const projects = r.results.find(domain => domain.q === 'TRK');
        return projects ? projects.items : [];
      })
      .then(projects =>
        projects.map(project => ({
          value: project.key,
          label: project.name,
          organization: project.organization
        }))
      )
      .then(options => {
        cb(null, { options });
      });
  };

  handleAddProject = selected => {
    const project = {
      key: selected.value,
      name: selected.label,
      organization: selected.organization
    };
    this.setState({
      addedProjects: [...this.state.addedProjects, project]
    });
  };

  renderOption = option => {
    return (
      <span>
        <Organization organizationKey={option.organization} link={false} />
        <strong>{option.label}</strong>
      </span>
    );
  };

  render() {
    const allProjects = [...this.props.projects, ...this.state.addedProjects];

    return (
      <section>
        <h2 className="spacer-bottom">
          {translate('my_profile.per_project_notifications.title')}
        </h2>

        {allProjects.length === 0 &&
          <div className="note">
            {translate('my_account.no_project_notifications')}
          </div>}

        {allProjects.map(project => <ProjectNotifications key={project.key} project={project} />)}

        <div className="spacer-top panel bg-muted">
          <span className="text-middle spacer-right">
            Set notifications for:
          </span>
          <Select.Async
            autoload={false}
            cache={false}
            name="new_project"
            style={{ width: '300px' }}
            loadOptions={this.loadOptions}
            minimumInput={2}
            optionRenderer={this.renderOption}
            onChange={this.handleAddProject}
            placeholder="Search Project"
          />
        </div>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  projects: getProjectsWithNotifications(state)
});

export default connect(mapStateToProps)(Projects);

export const UnconnectedProjects = Projects;
