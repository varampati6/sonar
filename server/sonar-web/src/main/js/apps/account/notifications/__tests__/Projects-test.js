/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import { UnconnectedProjects } from '../Projects';

const projects = [{ key: 'foo', name: 'Foo' }, { key: 'bar', name: 'Bar' }];

const newProject = { key: 'qux', name: 'Qux' };

it('should render projects', () => {
  const wrapper = shallow(<UnconnectedProjects projects={projects} />);
  expect(wrapper).toMatchSnapshot();

  // let's add a new project
  wrapper.setState({ addedProjects: [newProject] });
  expect(wrapper).toMatchSnapshot();

  // let's say we saved it, so it's passed back in `props`
  wrapper.setProps({ projects: [...projects, newProject] });
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.state()).toMatchSnapshot();
});
