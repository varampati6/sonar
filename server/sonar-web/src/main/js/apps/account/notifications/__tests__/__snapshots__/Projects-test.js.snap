// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render projects 1`] = `
<section>
  <h2
    className="spacer-bottom"
  >
    my_profile.per_project_notifications.title
  </h2>
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "foo",
        "name": "Foo",
      }
    }
  />
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "bar",
        "name": "Bar",
      }
    }
  />
  <div
    className="spacer-top panel bg-muted"
  >
    <span
      className="text-middle spacer-right"
    >
      Set notifications for:
    </span>
    <Async
      autoload={false}
      cache={false}
      ignoreAccents={true}
      ignoreCase={true}
      loadOptions={[Function]}
      loadingPlaceholder="Loading..."
      minimumInput={2}
      name="new_project"
      onChange={[Function]}
      optionRenderer={[Function]}
      options={Array []}
      placeholder="Search Project"
      searchPromptText="Type to search"
      style={
        Object {
          "width": "300px",
        }
      }
    />
  </div>
</section>
`;

exports[`should render projects 2`] = `
<section>
  <h2
    className="spacer-bottom"
  >
    my_profile.per_project_notifications.title
  </h2>
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "foo",
        "name": "Foo",
      }
    }
  />
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "bar",
        "name": "Bar",
      }
    }
  />
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "qux",
        "name": "Qux",
      }
    }
  />
  <div
    className="spacer-top panel bg-muted"
  >
    <span
      className="text-middle spacer-right"
    >
      Set notifications for:
    </span>
    <Async
      autoload={false}
      cache={false}
      ignoreAccents={true}
      ignoreCase={true}
      loadOptions={[Function]}
      loadingPlaceholder="Loading..."
      minimumInput={2}
      name="new_project"
      onChange={[Function]}
      optionRenderer={[Function]}
      options={Array []}
      placeholder="Search Project"
      searchPromptText="Type to search"
      style={
        Object {
          "width": "300px",
        }
      }
    />
  </div>
</section>
`;

exports[`should render projects 3`] = `
<section>
  <h2
    className="spacer-bottom"
  >
    my_profile.per_project_notifications.title
  </h2>
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "foo",
        "name": "Foo",
      }
    }
  />
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "bar",
        "name": "Bar",
      }
    }
  />
  <Connect(ProjectNotifications)
    project={
      Object {
        "key": "qux",
        "name": "Qux",
      }
    }
  />
  <div
    className="spacer-top panel bg-muted"
  >
    <span
      className="text-middle spacer-right"
    >
      Set notifications for:
    </span>
    <Async
      autoload={false}
      cache={false}
      ignoreAccents={true}
      ignoreCase={true}
      loadOptions={[Function]}
      loadingPlaceholder="Loading..."
      minimumInput={2}
      name="new_project"
      onChange={[Function]}
      optionRenderer={[Function]}
      options={Array []}
      placeholder="Search Project"
      searchPromptText="Type to search"
      style={
        Object {
          "width": "300px",
        }
      }
    />
  </div>
</section>
`;

exports[`should render projects 4`] = `
Object {
  "addedProjects": Array [],
}
`;
